### PR TITLE
EnclosureModule for HPE OneView

### PR DESCRIPTION
**Pull request by fgbulsoni**
##### SUMMARY
Added new oneview_enclosure module for retrieving [HPE OneView Enclosure](http://h17007.www1.hpe.com/docs/enterprise/servers/oneview3.0/cic-api/en/api-docs/current/index.html#rest/enclosures) resource and unit tests.

Related issue for more information: [HPE OneView support](https://github.com/ansible/ansible/issues/28354)

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
`oneview_enclosure`

##### ANSIBLE VERSION
```
ansible 2.5.0 (hpe-oneview/enclosure cfa05bc843) last updated 2017/09/15 19:00:33 (GMT +000)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/vagrant/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/vagrant/dev/ansible/lib/ansible
  executable location = /home/vagrant/dev/ansible/bin/ansible
  python version = 2.7.9 (default, Aug 30 2017, 01:33:10) [GCC 4.8.4]
```

##### ADDITIONAL INFORMATION
Example of usage:
```yaml
- name: Ensure that an Enclosure is present using the default configuration
  oneview_enclosure:
    config: /etc/oneview/oneview_config.json
    state: present
    data:
      enclosureGroupUri : /rest/enclosure-groups/my-uri
      hostname : 172.18.1.13
      username : admin
      password : password
      name: Test-Enclosure
      licensingIntent : OneView

- name: Updates the enclosure to have a name of "Test-Enclosure-Renamed".
  oneview_enclosure:
    config: /etc/oneview/oneview_config.json
    state: present
    data:
      name: Test-Enclosure
      newName : Test-Enclosure-Renamed

- name: Reconfigure the enclosure "Test-Enclosure"
  oneview_enclosure:
    config: /etc/oneview/oneview_config.json
    state: reconfigured
    data:
      name: Test-Enclosure

- name: Ensure that enclosure is absent
  oneview_enclosure:
    config: /etc/oneview/oneview_config.json
    state: absent
    data:
      name: Test-Enclosure

- name: Ensure that an enclosure is refreshed
  oneview_enclosure:
    config: /etc/oneview/oneview_config.json
    state: refreshed
    data:
      name: Test-Enclosure
      refreshState: Refreshing

- name: Set the calibrated max power of an unmanaged or unsupported enclosure
  oneview_enclosure:
    config: /etc/oneview/oneview_config.json
    state: present
    data:
      name: Test-Enclosure
      calibratedMaxPower: 1700

- name: Set the appliance bay power state on
  oneview_enclosure:
    config: /etc/oneview/oneview_config.json
    state: appliance_bays_powered_on
    data:
      name: Test-Enclosure
      bayNumber: 1

- name: Set the appliance UID state on
  oneview_enclosure:
    config: /etc/oneview/oneview_config.json
    state: uid_on
    data:
      name: Test-Enclosure

- name: Set the appliance UID state off
  oneview_enclosure:
    config: /etc/oneview/oneview_config.json
    state: uid_off
    data:
      name: Test-Enclosure

- name: Set the UID for the Synergy Frame Link Module state on
  oneview_enclosure:
    config: /etc/oneview/oneview_config.json
    state: manager_bays_uid_on
    data:
      name: Test-Enclosure
      bayNumber: 1

- name: Set the UID for the Synergy Frame Link Module state off
  oneview_enclosure:
    config: /etc/oneview/oneview_config.json
    state: manager_bays_uid_off
    data:
      name: Test-Enclosure
      bayNumber: 1

- name: E-Fuse the Synergy Frame Link Module bay 1
  oneview_enclosure:
    config: /etc/oneview/oneview_config.json
    state: manager_bays_power_state_e_fuse
    data:
      name: Test-Enclosure
      bayNumber: 1

- name: Reset the Synergy Frame Link Module bay 1
  oneview_enclosure:
    config: /etc/oneview/oneview_config.json
    state: manager_bays_power_state_reset
    data:
      name: Test-Enclosure
      bayNumber: 1

- name: E-Fuse the appliance bay 1
  oneview_enclosure:
    config: /etc/oneview/oneview_config.json
    state: appliance_bays_power_state_e_fuse
    data:
      name: Test-Enclosure
      bayNumber: 1

- name: E-Fuse the device bay 10
  oneview_enclosure:
    config: /etc/oneview/oneview_config.json
    state: device_bays_power_state_e_fuse
    data:
      name: Test-Enclosure
      bayNumber: 10

- name: Reset the device bay 8
  oneview_enclosure:
    config: /etc/oneview/oneview_config.json
    state: device_bays_power_state_reset
    data:
      name: Test-Enclosure
      bayNumber: 8

- name: E-Fuse the IC bay 3
  oneview_enclosure:
    config: /etc/oneview/oneview_config.json
    state: interconnect_bays_power_state_e_fuse
    data:
      name: Test-Enclosure
      bayNumber: 3

- name: Set the active Synergy Frame Link Module on bay 2
  oneview_enclosure:
    config: /etc/oneview/oneview_config.json
    state: manager_bays_role_active
    data:
      name: Test-Enclosure
      bayNumber: 2

- name: Release IPv4 address in the device bay 2
  oneview_enclosure:
    config: /etc/oneview/oneview_config.json
    state: device_bays_ipv4_removed
    data:
      name: Test-Enclosure
      bayNumber: 2

- name: Release IPv4 address in the interconnect bay 2
  oneview_enclosure:
    config: /etc/oneview/oneview_config.json
    state: interconnect_bays_ipv4_removed
    data:
      name: Test-Enclosure
      bayNumber: 2

- name: Set the supportDataCollectionState for the enclosure
  oneview_enclosure:
    config: /etc/oneview/oneview_config.json
    state: support_data_collection_set
    data:
      name: Test-Enclosure
      supportDataCollectionState: PendingCollection

- name: Ensure that the Enclosure is present and is inserted in the desired scopes
  oneview_enclosure:
    config: /etc/oneview/oneview_config.json
    state: present
    data:
      name: Test-Enclosure
      scopeUris:
        - /rest/scopes/00SC123456
        - /rest/scopes/01SC123456
```

### CODE REVIEW

File: lib/ansible/modules/remote_management/oneview/oneview_enclosure.py
Comment (tmiotto) [inline lib/ansible/modules/remote_management/oneview/oneview_enclosure.py]: These messages seem to be in the present, shouldn't they have to be in past tense?

**Comment by tmiotto** [comment]
shipit

**Comment by fgbulsoni** [comment]
@abadger , @dagwieers & any who feel like it, the PR is ready to be merged as far as I'm concerned. Please review it whenever possible. :octocat: 

Also, this is probably one of the biggest modules for oneview, so do feel free to suggest improvements.

**Comment by fgbulsoni** [comment]
ready_for_review

File: lib/ansible/modules/remote_management/oneview/oneview_enclosure.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/oneview/oneview_enclosure.py]: Blank line before this one, and format like:
```python
# Copyright: (c) 2016-2017, Hewlett Packard Enterprise Development LP
```

**Review by dagwieers — CHANGES_REQUESTED** [review]

**Comment by fgbulsoni** [comment]
Hey @dagwieers , thanks for the comment, I have fixed the PR accordingly. 
I think it is now ready_for_review , let me know if there's anything else or feel free to merge.

**Review by dagwieers — APPROVED** [review]

**Review by dagwieers — COMMENTED** [review]
Hmm, after closer inspection why there is such a long description I think the interface is weird. The state options are not all states, it rather seems a selector. That is quite confusing...

**Comment by dagwieers** [comment]
bot_status

**Comment by fgbulsoni** [comment]
> Hmm, after closer inspection why there is such a long description I think the interface is weird. The state options are not all states, it rather seems a selector. That is quite confusing...

Hmm, yeah, this PR has a *lot* of states. 
Let me know if any of those particularly scream out a particular need to be changed in name or something similar. 
This resource and the server profile ones are likely the biggest resources for oneview.

**Comment by tmiotto** [comment]
shipit

**Comment by adriane-cardozo** [comment]
shipit

**Comment by fgbulsoni** [comment]
bot_status

**Comment by fgbulsoni** [comment]
> maintainer_shipits (module maintainers): 0
> community_shipits (namespace maintainers): 3
> ansible_shipits (core team members): 0

Hey @dagwieers , I've tried the multiple `shipit`s as we've talked about on [#28844](https://github.com/ansible/ansible/pull/28844#issuecomment-332856865) and the PRs do not appear to be merging either way. 

So, I think we still require an approval or shipit from someone listed as `ansible_shipits (core team members)`. Let me know how to proceed. :octocat:

File: lib/ansible/modules/remote_management/oneview/oneview_enclosure.py
Comment (bcoca) [inline lib/ansible/modules/remote_management/oneview/oneview_enclosure.py:37]: most of these should not be states but individual toggles.

**Review by bcoca — CHANGES_REQUESTED** [review]

**Comment by fgbulsoni** [comment]
This PR requires some changes which will most likely be performed by another member of the team, so for now I'm closing it until it has been refactored/is good to go.

Thanks guys :octocat:

**Pull request closed without merging**
